solr/core/src/test/org/apache/solr/TestDistributedGrouping.java

```diff
@@ -425,6 +426,65 @@ public void test() throws Exception {
     
     //Debug
     simpleQuery("q", "*:*", "rows", 10, "fl", "id," + i1, "group", "true", "group.field", i1, "debug", "true");
+    doTestGroupSkipSecondStep();
+  }
+
+  /*
+    SOLR-11831, test skipping the second grouping step if the query only retrieves on document per group
+   */
+  private void doTestGroupSkipSecondStep() throws Exception {
+    ignoreException(GroupParams.GROUP_SKIP_DISTRIBUTED_SECOND); // don't print stack trace for exception raised by group.skip.second.step
+    // Ignore numFound if group.skip.second.step is enabled because the number of documents per group will not be computed (will default to 1)
+    handle.put("numFound", SKIP);
+    query("q", "{!func}id_i1", "rows", 3, "group.skip.second.step", true, "group.limit", 1,  "fl",  "id," + i1, "group", "true",
+        "group.field", i1);
+    query("q", "kings", "group.skip.second.step", true, "fl", "id," + i1, "group", "true", "group.field", i1);
+    query("q", "{!func}id_i1", "rows", 3, "group.skip.second.step", true,  "fl",  "id," + i1, "group", "true",
+        "group.field", i1);
+    query("q", "1234doesnotmatchanything1234", "group.skip.second.step", true, "fl", "id," + i1, "group", "true", "group.field", i1);
+
+    ignoreException("Illegal grouping specification");
+    // ngroups will return the corrent results, the problem is that numFound for each group might be wrong in case of multishard setting - but there is no way to
+    // enable/disable it.
+    //assertSimpleQueryThrows("q", "{!func}id_i1", "group.skip.second.step", true, "fl", "id," + i1, "group", "true", "group.field", i1, "group.ngroups", true);
+    assertSimpleQueryThrows("q", "{!func}id", "group.skip.second.step", true, "fl", "id," + i1, "group", "true", "group.field", i1, "group.limit", 5);
+    assertSimpleQueryThrows("q", "{!func}id_i1", "group.skip.second.step", true, "fl", "id," + i1, "group", "true", "group.field", i1, "group.limit", 0);
+    // group sorted in a different way should fail
```

Comment:
These two tests after the `group sorted in a different way should fail` comment are interesting:
* the comment suggests that failure is expected because of `sort/group.sort` parameters but wouldn't it also fail because of the `group.limit=0` i.e. to truely test about sorting the group.limit should be absent (default is 1) or explicitly set to 1.
* eliminating the group.limit would give the queries below but shouldn't the first of the two queries pass?
  * query 1: if sort is present and group.sort is absent then sort would also be used for group.sort -- it should pass?
  * query 2: sort is absent and so the default is `sort=score` (I think) of which the `group.sort` is not a prefix match -- it should fail.
```
assertSimpleQueryThrows(...,"sort", i1+" desc");
assertSimpleQueryThrows(..., "group.sort", i1+" desc");
```

Reply:
Following on from the above reasoning, test coverage for `sort` and `group.sort` interaction with `group.skip.second.step` could include something like

```
query("q", "{!func}id_i1", "group.skip.second.step", true, "fl", "id," + i1, "group", "true", "group.field", i1, "group.limit", 1,
      "sort", i1+" desc");
query("q", "{!func}id_i1", "group.skip.second.step", true, "fl", "id," + i1, "group", "true", "group.field", i1, "group.limit", 1,
      "sort", i1+" desc", "group.sort", i1+" desc");
```

i.e. `group.sort` being absent is fine and `group.sort` matching `sort` is fine too. These should pass, I think, no?

https://github.com/cpoerschke/lucene-solr/commit/3690cb3d4ed537546d2d876eb00dcc0fb735a557 has those two tests commented out because mysteriously most of the time they fail (but part of the time they succeed!). Will add failure example output.

Reply:
I think I found an answer to the enigma of these tests failing, the main observation was that *the tests were occasionally failing also disabling `group.skip.second.step`*. 
The issue are two documents in the collection with the same `i1` field:

```
doc1: id_i1=19, i1=232
doc2: id_i1=18, i1=232
```

If we have only one shard result will be 18 (**the sort value is the same** and then on a single shard documents will be sorted by the internal `doc_id` and the first one inserted (lower `doc_id`) wins.)

But on two or more shards, if the two documents are on different shards, one shard will return the `doc1` (score 232), and the other `doc2` (score 232) - and the merge won't have information about the internal `doc_id` of the two docs (also if you are particularly unlucky nothing is stopping `doc_id` from being the same ;) ). 

To solve I added the solr logical key ( (`id_i1`) as  secondary solr field and now  those tests are fine.